File: traceforge/sdk/python/traceforge/_emit.py
# SPDX-License-Identifier: MIT
from __future__ import annotations
import json
import logging
import os
import urllib.request
from typing import Sequence

from ._span import Span

logger = logging.getLogger(__name__)

_DEFAULT_ENDPOINT = "http://localhost:8080/v1/spans"


def emit_spans(spans: Sequence[Span], endpoint: str | None = None) -> None:
    """POST spans to the TraceForge collector. Fire-and-forget; swallows errors."""
    url = endpoint or os.getenv("TRACEFORGE_ENDPOINT", _DEFAULT_ENDPOINT)
    payload = json.dumps([s.to_dict() for s in spans]).encode()
    req = urllib.request.Request(
        url,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        with urllib.request.urlopen(req, timeout=2) as resp:
            if resp.status != 200:
                logger.warning("traceforge: collector returned %d", resp.status)
    except Exception as exc:
        logger.debug("traceforge: emit failed (%s) — continuing", exc)

File: traceforge/sdk/go/traceforge/emit.go
// SPDX-License-Identifier: MIT
package traceforge

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/twmb/franz-go/pkg/kgo"
)

var (
	httpClient = &http.Client{Timeout: 2 * time.Second}
	kafkaOnce  sync.Once
	kafkaClient *kgo.Client
)

func collectorURL() string {
	u := os.Getenv("TRACEFORGE_COLLECTOR_URL")
	if u == "" {
		u = "http://localhost:8080/v1/spans"
	}
	return u
}

func kafkaBrokers() []string {
	raw := os.Getenv("TRACEFORGE_KAFKA_BROKERS")
	if raw == "" {
		return nil
	}
	return strings.Split(raw, ",")
}

func kafkaTopic() string {
	t := os.Getenv("TRACEFORGE_KAFKA_TOPIC")
	if t == "" {
		t = "agent-spans"
	}
	return t
}

// emit sends span over HTTP to the collector and, if Kafka is configured, also
// produces to the agent-spans topic. Both sends are fire-and-forget: failures
// are silently dropped to avoid blocking the caller.
func emit(ctx context.Context, s *Span) {
	payload, err := json.Marshal(s)
	if err != nil {
		return
	}
	go sendHTTP(payload)
	if brokers := kafkaBrokers(); len(brokers) > 0 {
		go sendKafka(ctx, brokers, payload)
	}
}

func sendHTTP(payload []byte) {
	req, err := http.NewRequest(http.MethodPost, collectorURL(), bytes.NewReader(payload))
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := httpClient.Do(req)
	if err != nil {
		return
	}
	resp.Body.Close()
}

func sendKafka(ctx context.Context, brokers []string, payload []byte) {
	kafkaOnce.Do(func() {
		cl, err := kgo.NewClient(
			kgo.SeedBrokers(brokers...),
			kgo.DefaultProduceTopic(kafkaTopic()),
		)
		if err == nil {
			kafkaClient = cl
		}
	})
	if kafkaClient == nil {
		return
	}
	kafkaClient.ProduceSync(ctx, &kgo.Record{Value: payload})
}
